### PR TITLE
Feature: use BPF filter to only get bonjour traffic

### PR DESCRIPTION
**Pull request by remicalixte**
Greatly improve CPU usage #20.
Edit: actually this does not improve performance but still makes the code a lot cleaner.
See #24 for performance improvements.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -43,10 +43,17 @@ func main() {
 	}
 	brMACAddress := intf.HardwareAddr
 
+	// Filter tagged bonjour traffic
+	filterTemplate := "not (ether src %s) and vlan and dst net (224.0.0.251 or ff02::fb) and udp dst port 5353"
```
[the remark this answers is not in the document]
Comment (remicalixte) [reply main.go]: Done

**Comment by L3Nerd** [comment]
@remicalixte I think you are wrong :-)
Did a checkout of your PR, cause of high cpu-usage on my homeserver (constant network-traffic on that interface), your branch got the process from ~20% to <1%.

**Comment by Gandem** [comment]
Let's get this PR merged then ! @remicalixte mind rebasing this on top of latest `master`, the CI is broken right now, and there is a couple of commits in master that should fix this

**Comment by remicalixte** [comment]
@L3Nerd Great ! Let's get this merge then

**Comment by remicalixte** [comment]
Alright the tests are fixed now

File: packet_test.go [hunk 2]
```diff
@@ -151,16 +150,16 @@ func TestParseIPLayer(t *testing.T) {
 	isIPv4 := true
 	ipv4Packet := gopacket.NewPacket(createMockmDNSPacket(isIPv4, true), decoder, options)
 
-	computedIPv4, computedIsIPv6 := parseIPLayer(ipv4Packet)
-	if !reflect.DeepEqual(dstIPv4Test, computedIPv4) || (computedIsIPv6 == isIPv4) {
+	computedIsIPv6 := parseIPLayer(ipv4Packet)
+	if computedIsIPv6 == isIPv4 {
```
Comment (Gandem) [inline packet_test.go]: Let's replace `isIPv4` with `true` to make this test case a bit more readable

File: packet_test.go [hunk 3]
```diff
@@ -151,16 +150,16 @@ func TestParseIPLayer(t *testing.T) {
...
 		t.Error("Error in parseIPLayer() for IPv4 addresses")
 	}
 
 	isIPv4 = false
 	ipv6Packet := gopacket.NewPacket(createMockmDNSPacket(isIPv4, true), decoder, options)
 
-	computedIPv6, computedIsIPv6 := parseIPLayer(ipv6Packet)
-	if !reflect.DeepEqual(dstIPv6Test, computedIPv6) || (computedIsIPv6 == isIPv4) {
+	computedIsIPv6 = parseIPLayer(ipv6Packet)
+	if computedIsIPv6 == isIPv4 {
```
Comment (Gandem) [inline packet_test.go]: Same here

File: packet_test.go [hunk 4]
```diff
@@ -251,13 +245,13 @@ func areBonjourPacketsEqual(a, b bonjourPacket) (areEqual bool) {
 	// While comparing Bonjour packets, we do not want to compare packets entirely.
 	// In particular, packet.metadata may be slightly different, we do not need them to be the same.
 	// So we only compare the layers part of the packets.
-	areEqual = areEqual && reflect.DeepEqual(a.packet.Layers(), b.packet.Layers())
+	areEqual = areEqual && reflect.DeepEqual(a.packet.Layers()[2], b.packet.Layers()[2])
```
Comment (Gandem) [inline packet_test.go]: What's the reasoning behind this change? Should we also edit the comment?

Comment (remicalixte) [reply packet_test.go]: Mmh sorry, leftover from debugging

File: packet_test.go (hunk 2 above)
Comment (remicalixte) [reply packet_test.go]: I'm still not a big fan of this. I would like to add `assert` to the tests to make them even clearer in the future

**Comment by oxlay** [comment]
Nice!! 🎉

**Pull request merged**